Added UpgradeModuleGadget for vehicles upgrades.

VELD-Dev opened the pull request:
### Changes made in this pull request

  - Added a gadget for vehicle upgrades, still in work but can do the job actually
  - Added the possibility to make delegates for many events of modules:
    - `WithOnUpgradeAdded(Action)` - The action takes many arguments, see definitions.
    - `WithOnUpgradeRemoved(Action)` - Same there
    - `WithOnUpgradeUsed(Action)` - There too
    - `WithOnUpgradeToggled(Action)` - There too, but nota bene: this even occurs both when the module is on and off. That's why the delegate has a `bool toggled` arugment, `toggled=true=turned on`

### Breaking changes

  - All the vehicles have separate patches, for many reason, but first is security. There's the patches for `Vehicle` class, for `SeaMoth`, for `Exosuit` (BZ and SN1), `SeaTruck` and SnowFox `Hoverbike`.

VELD-Dev committed:
Made the ModuleGadget thingy. Ready for PR ?

VELD-Dev committed:
Now ready for PR, test may be needed, if no push in 10 minutes, then it is a valid build.

VELD-Dev committed:
okay fixed something, now ready for PR

VELD-Dev committed:
changed the name of the module gadget to UpgradeModuleGadget

VELD-Dev committed:
Corrected a possible bug (latest)

VELD-Dev committed:
Merge branch 'SubnauticaModding:master' into master

VELD-Dev commented on Nautilus/Assets/Gadgets/EquipmentGadget.cs:
```diff
@@ -20,6 +20,7 @@ public class EquipmentGadget : Gadget
     public QuickSlotType QuickSlotType { get; set; }
 
 
+
```
These spaces are for readability, I think that spacing a little some parts is sometimes better for readability

VELD-Dev commented:
> ⚠️ A patch MUST be done on user side to `OnUpgradeModuleUse` and probably on `OnUpgradeModuleCharge` ! May be a good idea to include it into the helper as `.WithOnUpgradeUse(Action<Vehicle instance, double maxCharge, double energyCost> delegate)`, `.WithOnUpgradeAdded(Action<Vehicle instance> delegate)` and `.WithOnUpgradeRemoved(Action<Vehicle instance> delegate)`

I thought about making a kind of transpilers so we can add delegates and our actions on the functions for the requested changes on top of the original, for example for any basic upgrade we should be making some default actions if it is having a `.WithHullUpgrade(100f)`, if they use `.WithOnUpgradeAdded(Action<Vehicle>)` and that they play a sound or whatever, the hull is already added to the ship, and THEN it plays the sound, well we can do that for hull, depth, and for custom actions, just use the `.WithOn[insert action here](Action<Vehicle, double, double>)`  
  
Example:
```cs
public static void AddMyModule() 
{
    Info = PrefabInfo
        .WithTechType(classId: "ExampleItem", displayName: "Example module")
        .WithSizeInInventory(new Vector2(1, 1));

    CustomPrefab cf = new(Info);
    CloneTemplate clone = new(Info, TechType.UpgradeModuleHull1);
    
    cf.SetGameObject(clone)
    cf.SetUpgradeModule(EquipmentType.SeamothModule)
        .WithQuickSlotType(QuickSlotType.Chargeable)
        .WithEnergyCost(1f)
        .WithMaxCharge(10f)
        .WithOnUpgradeUse((Seamoth seamoth, double maxCharge, double energyCost) => {
            Logger.LogInfo("Hey! I have been used!")
        });
}
```
Meanwhile a transpiler would be running this callback:
```cs
public static void OnUpgradeUseCallback(Vehicle __instance)
{
    foreach(CustomPrefab CustomModule in ModulesHandler.CustomModules)
    {
        // Do our actions here, and invoke the callback. For hull it would be Seamoth.armor = CustomModule.hull,
        // and then the user runs his action, but we should be automating that.
    }
}
```
We can also detect the module type and transpile only on the concerned vehicle (SeamothModule on SeaMoth, ExosuitModule on Exosuit, etc...)

32Kallies commented on Nautilus/Assets/Gadgets/GadgetExtensions.cs:
```diff
@@ -136,6 +136,27 @@ public static EquipmentGadget SetEquipment(this ICustomPrefab customPrefab, Equi
+    /// <param name="equipmentType">The type of module slot this item can fit into.<br/>
+    /// If you plan to use anything that is not a module (Tank, Chip), except ExosuitArm,<br/>
+    /// you better use <see cref="SetEquipment(ICustomPrefab, EquipmentType)"/> instead.</param>
+    /// <returns>An instance to the created <see cref="UpgradeModuleGadget"/> to continue the equipment settings on.</returns>
```
Seems like the documentation needs to be edited

VELD-Dev replied:
What should I put instead ?

VELD-Dev replied:
Nah Ill just remove that and add it to Equipment Gadget

32Kallies commented on Nautilus/Assets/Gadgets/UpgradeModuleGadget.cs:
```diff
@@ -0,0 +1,114 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Unnecessary using statements

VELD-Dev replied:
i can do a global cleanup if you want XDD

32Kallies commented on Nautilus/Assets/Gadgets/UpgradeModuleGadget.cs:
```diff
@@ -0,0 +1,114 @@
+namespace Nautilus.Assets.Gadgets;
+
+/// <summary>
+/// Represents a vehicle module (or upgrade) gadget
```
Add period

VELD-Dev replied:
Allright i'm doing this

32Kallies commented on Nautilus/Assets/Gadgets/UpgradeModuleGadget.cs:
```diff
@@ -0,0 +1,114 @@
+
+    /// <summary>
+    /// Max charge of this item.
+    /// Should apply to modules of vehicles and to chargeable items.
```
Perhaps elaborate just a little more on what this does, or give an example of where it is used in the base game

32Kallies commented on Nautilus/Assets/Gadgets/UpgradeModuleGadget.cs:
```diff
@@ -0,0 +1,114 @@
+            return;
+        }
+
+        CraftDataHandler.SetEquipmentType(prefab.Info.TechType, ModuleType);
```
I don't think this is a good idea, since we already have the EquipmentGadget.

32Kallies commented:
Besides the review comments I sent, it seems good to me. The main thing I'm really not sure on is why we need to include the ModuleType property when we already have the EquipmentGadget.

32Kallies commented:
Of course it still needs the patches to have most of its behaviour and... that's quite a big task. I don't have a mod project to test out those features with either.

VELD-Dev commented:
> Of course it still needs the patches to have most of its behaviour and... that's quite a big task. I don't have a mod project to test out those features with either.

Yeah sure, I'll do that but I must learn more about how I could do with the transpilers you know, because there is a lot of things I'm unsure such as the callbacks and all that, I got an idea but it would be easier to show on Discord or in vocal not gonna lie.

VELD-Dev commented:
> Besides the review comments I sent, it seems good to me. The main thing I'm really not sure on is why we need to include the ModuleType property when we already have the EquipmentGadget.

I did that so we don't need to use .SetEquipment since it's not really an equipment

VELD-Dev commented:
Do not merge my PR yet, I'm working on those functionalities, I'm making transpilers to avoid post or pre fixes and to have a better compatibility

32Kallies commented:
Where exactly would a postfix not work?

32Kallies commented:
You might need to rename the file again (sorry) to make it clear that it only is used for “vehicles” (which is a categorization that does not encompass the cyclops).

32Kallies commented on Nautilus/Assets/Gadgets/UpgradeModuleGadget.cs:
```diff
@@ -0,0 +1,114 @@
+/// <summary>
+/// Represents a vehicle module (or upgrade) gadget
+/// </summary>
+public class UpgradeModuleGadget : Gadget
```
Perhaps rename to "VehicleUpgradeModuleGadget", since the game considers "Vehicles" to be everything except the Cyclops

VELD-Dev committed:
Added an insane quantity of functionality, and it's only the beginning. The most advanced part is for Subnautica 1, for Below Zero I still have a loooot of work, and SN1 too but BZ more...

VELD-Dev committed:
*tried* to add an example upgrade module

VELD-Dev committed:
now the example mod should work

VELD-Dev committed:
Seamoth, Seatruck and Snowbike are now compatible with Upgrade Module Gadget. Almost finished the rest.

VELD-Dev committed:
Merge branch 'master' into master

VELD-Dev commented:
Nota bene: Even if I said in my last commits that it was compatible, I still have not done the transpilers for on use and on toggle delegates. It's important to note because a lot of work has to be done in fact.

VELD-Dev committed:
Cleaned all unnecessary usings, as Lee asked.

VELD-Dev committed:
Added SeaTruck On Use delegate, should be working at 100% but test is needed, if someone can make an example mod for that (BZ) it would be great, I don't have time lastly

VELD-Dev commented:
Tests required for the last fix, I added a `WithCooldown(double seconds)` that works only with SeaTruck modules now, I need a test...

VELD-Dev committed:
Added Seamoth OnUse patch for test with Example Mod.
The delegate for SeaMoth is now working holy shit.

VELD-Dev committed:
Merge remote-tracking branch 'upstream/master'

VELD-Dev committed:
Removed harmony debug attribute

VELD-Dev committed:
Added some stuff

VELD-Dev committed:
Tests are needed. Normally everything is working **EXCEPT HOVERBIKE OnUse, HOVERBIKE OnToggle and (maybe) SEATRUCK OnToggle.**
I am writing a MonoBehaviour for the Hoverbike and one for the SeaTruck (HoverbikeComplement and SeaTruckComplement, possibly renamed in the future) but I will add it in a different PR.

VELD-Dev committed:
Merge branch 'master' into master

VELD-Dev commented:
Tests are needed. The best way to have some is by merging.
So let's say that it's ready for merge. I will try to fix the bugs as quick as possible when it's released and we get bug reports.
While this, I'll working on HoverbikeCustomQuickSlot monobehaviour.

32Kallies committed:
Revert autoformatting & using statement reordering

32Kallies committed:
Umm... no

32Kallies committed:
What am I reverting?

32Kallies committed:
Remove Vehicle Upgrade mod csproj files

32Kallies committed:
Fix more odd oversights

32Kallies committed:
?

32Kallies committed:
Callback shenanigans

32Kallies committed:
Remove unused using statements

32Kallies committed:
Move all example code into VehicleUpgradesExample

32Kallies committed:
Merge remote-tracking branch 'upstream/master' into pr/391

VELD-Dev committed:
Potentially fixed seamoth crush depth not changing. Debug, troubleshoot and tests are required

32Kallies committed:
Merge remote-tracking branch 'upstream/master' into pr/391

32Kallies committed:
Merge branch 'master' of https://github.com/VELD-Dev/Nautilus into pr/391

VELD-Dev committed:
Okay well.. It's working !

VELD-Dev committed:
bring some additional changes requested by sir Metious.

32Kallies commented on Nautilus/Assets/Gadgets/GadgetExtensions.cs:
```diff
@@ -147,6 +148,31 @@ public static EquipmentGadget SetEquipment(this ICustomPrefab customPrefab, Equi
+    /// <summary>
+    /// Sets this item as a vehicle upgrade module. Cyclops upgrades are not supported by this function.
+    /// <para>If you're using this function, please do not use <see cref="SetEquipment(ICustomPrefab, EquipmentType)"/>,<br/>
+    /// it would interefere with this and possibly make the game crash or the mod not to work.</para>
```
Grammar nitpick: Change "the mod not to work" to "cause the mod to not work", or something similar.

However, calling SetEquipment wouldn't do either thing, I think, in fact it would do nothing at all since you are using the gadget extensions. Those are always safe.

The rest of these docs are good btw.

VELD-Dev replied:
Oh okay, I'll fix that thanks I was really not sure when I was writing :D

VELD-Dev replied:
Well, I must at least give an error if the item has no Equipment gadget...

32Kallies commented on Nautilus/Patchers/VehicleUpgradesPatcher.cs:
```diff
@@ -0,0 +1,522 @@
+                    if (techType == TechType.None)
+                        break;
+
+                    if (!SeamothUpgradeModules.TryGetValue(techType, out prefab))
```
Causes a compiler error on the Below Zero build target.

VELD-Dev replied:
Ah yeah I must put that in regions my bad

32Kallies commented on Example mod/VehicleUpgradesExample.cs:
This class also causes compiler errors when building for Below Zero. Perhaps just exclude the entire file from BZ until you have a Seatruck module example.

VELD-Dev replied:
Yeah I'll do that

32Kallies reviewed (changes requested):
There's too much to review in one night, but it all seems good besides the few things I pointed out. I will not waste hours reading every last line of code and documentation, and instead I will trust that it is clean and functional. It did indeed work in my tests, and I see no glaring issues after glancing through it all.

VELD-Dev committed:
Edited the primary problems for merge I guess

VELD-Dev committed:
removed the "configurable and changeable values" to be in a next PR that I will probably do very quickly so at least it's done

32Kallies committed:
Removed unnecessary using statement

32Kallies committed:
Update VehicleUpgradesExample.cs

VELD-Dev committed:
Merge branch 'SubnauticaModding:master' into master

VELD-Dev committed:
added a documentation page

32Kallies commented on Nautilus/Documentation/tutorials/toc.yml:
```diff
@@ -11,6 +11,8 @@
       href: story-goals.md
     - name: Databank entries
       href: databank-entries.md
+    - name: Custom vehicle upgrade module
```
"module" -> "modules" for consistency (all others are plural)

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+# Using the Vehicle Upgrade Module Gadget
+
+Vehicles modules are often hard to mod, and when multiple mods patch a function of a vehicle it can make severe conflicts leading to one of both mods to not work.  
```
"leading to one of both mods to not work" sounds a bit unnatural, maybe change it to something like "which can cause one or both mods to break"

VELD-Dev replied:
I'm changing that

VELD-Dev committed:
finished to update the doc

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+
+This brand new gadget was made to be compatible with both Below Zero and Subnautica 1, some functions differ between Nautilus for SN1 and Nautilus for BZ.
+
+## Possible actions on completion
```
Might want to rename the title here, to something like "Possibilities of custom vehicle upgrade modules"... idk, I don't have any great ideas

Also by any chance did you use the story goals guide as a basis?

VELD-Dev replied:
Yes I did take the story goal as a basis since I did not see anything like this on the site

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+
+## Possible actions on completion
+
+- Set the new crush depth of the vehicle when the upgrade is equipped. You can choose wether the depth is absolute or added to the default crush depth of the vehicle.
```
"wether" seems to be a typo

VELD-Dev replied:
It is ?

32Kallies replied:
Yeah, "whether"

VELD-Dev replied:
I think I will sleep after that lmao

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+## Possible actions on completion
+
+- Set the new crush depth of the vehicle when the upgrade is equipped. You can choose wether the depth is absolute or added to the default crush depth of the vehicle.
+- Set the max charge of the module if the module slot type is set to `QuickSlotType.SelectableChargeable` or `QuickSlotType.Chargeable`. The way to mesure the charge is still blurry.
```
"mesure" -> "measure"

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+
+Vehicles modules are often hard to mod, and when multiple mods patch a function of a vehicle it can make severe conflicts leading to one of both mods to not work.  
+The Vehicle Upgrade Module Gadget is made to make your life of modder as easy as possible when it comes to vehicle modules.  
+
```
Inconsistent spacing with paragraphs. I prefer more spacing over none.

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+
+## Chargeable self-destruct module
+
+Admitting we've already done the prefab info and the custom prefab, let's directly work only with the upgrade module gadget.
```
Odd choice of "admitting" here. It makes sense, but at the same time it really doesn't...

VELD-Dev replied:
What should I say ?

32Kallies replied:
I'm struggling with figuring out how to word it ngl. "Since"?

VELD-Dev replied:
Assuming ?

32Kallies replied:
Well there is no assumption happening here, since the code examples above already mention PrefabInfo.

VELD-Dev replied:
in fact, it's like if there was a reset between the two examples, they are independent, i just tell to assume that we've already done a prefab info and custom prefab for this new example I guess

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+        continue;
+    }
+    instance.liveMixin.Kill(DamageType.Explosive);
+}
```
Funny example lol, but it seems like it would crash the game. Should be a coroutine, with a return type of IEnumerator, called like `UWE.CoroutineHost.StartCoroutine(EngageSelfDestruct)`, and instead of `continue`, put `yield return null`.

Edit: You also need to cache the time that the cooldown began, so like

```cs
float startTime = Time.time;
while(Time.time < (startTime + countdown))
// ...
```

VELD-Dev replied:
So in an IEnumerable ?

VELD-Dev replied:
Ah no, as a CoroutineTask

VELD-Dev replied:
i forgor 💀

32Kallies replied:
Nope, the IEnumerator return type is what Unity uses for [coroutines](https://docs.unity3d.com/Manual/Coroutines.html).

32Kallies replied:
Btw, the problem before was that you were waiting for Time.time to change, but it never has a chance to change because time can only continue AFTER that method reaches its end.

VELD-Dev replied:
then should be Unity's StartCoroutine ?

VELD-Dev replied:
> Btw, the problem before was that you were waiting for Time.time to change, but it never has a chance to change because time can only continue AFTER that method reaches its end.

Fixed

32Kallies replied:
`UWE.CoroutineHost.StartCoroutine(...)` is better here because it can be called from anywhere

VELD-Dev replied:
ok alright, so the type of the function must be a CoroutineTask or a CoroutineHost ?

32Kallies replied:
Neither, IEnumerator should be the return type because Unity expects a coroutine to have that

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
```diff
@@ -0,0 +1,137 @@
+    })
+    .WithOnModuleRemoved((Vehicle inst, int slotId) =>
+    {
+        Subttiles.Add("Self-destruct module uninstalled.");
```
Subttiles

VELD-Dev replied:
it's very very late here lol

32Kallies commented on Nautilus/Documentation/tutorials/vehicle-module.md:
You need to add this to the overview.md file too

32Kallies reviewed (changes requested):
Overall the docs are all good, but here are some suggestions to keep it more clean. And also I noticed an issue with one of the code segments.

VELD-Dev committed:
This time it should be good

32Kallies commented:
recheck

VELD-Dev commented:
I have read the CLA Document and I hereby sign the CLA

VELD-Dev committed:
VehicleUpgradesPatcher fixed, upgrades are now working totally on Exosuit too !

VELD-Dev committed:
Cleaned unusded usings.

VELD-Dev committed:
Merge branch 'SubnauticaModding:master' into master

32Kallies commented on Nautilus/Patchers/VehicleUpgradesPatcher.cs:
```diff
@@ -35,28 +34,59 @@ internal static void Patch(Harmony harmony)
-        if (!prefab.TryGetGadget(out UpgradeModuleGadget moduleGadget))
-            return;
+            if (!prefab.TryGetGadget(out UpgradeModuleGadget moduleGadget))
+                return;
```
Could this be moved to the beginning of the patch?

VELD-Dev replied:
No, I'm making a distinction between ExosuitUpgradeModules and VehicleUpgradeModules

32Kallies commented on Nautilus/Patchers/VehicleUpgradesPatcher.cs:
```diff
@@ -133,59 +171,111 @@ private static void LazyInitialize(Vehicle __instance)
+
+        return matcher.InstructionEnumeration();
     }
+    */
```
Not very clean to leave commented out code unless it is almost done and you plan on picking it up very soon

VELD-Dev replied:
ooh yeah my bad i'll remove that

32Kallies commented on Nautilus/Patchers/VehicleUpgradesPatcher.cs:
```diff
@@ -133,59 +171,111 @@ private static void LazyInitialize(Vehicle __instance)
 
-            if (!prefab.TryGetGadget(out UpgradeModuleGadget moduleGadget))
-                return false;
+        if (!prefab.TryGetGadget(out moduleGadget))
```
Again, couldn't this be moved to the top?

VELD-Dev replied:
Nope, making distinction between ExosuitUpgradeModules and VehicleUpgradeModules

VELD-Dev committed:
Optimized code and removed comments.

VELD-Dev committed:
Merge branch 'master' of https://github.com/VELD-Dev/Nautilus

32Kallies committed:
Must be "yield return null"

32Kallies committed:
Fix typo ("POssibilities")

32Kallies committed:
Minor grammar changes
